### PR TITLE
LPS-135563 Fix placement of Balloon Editor panel

### PR DESCRIPTION
**Pull request by markocikos**
JIRA: https://issues.liferay.com/browse/LPS-135563

https://github.com/liferay/liferay-ckeditor/pull/164#issuecomment-825522824 is the most detailed summary of desired behavior for panel positioning I could find, along with discussion of challenges in implementation.

In this PR, we are fixing panel positioning, to reflect most if not all of above requirements:
- Selection in single row, either with double-click, keyboard or mouse drag, positions panel on top, centered on selection.
- Multiple row selection centers on end of selection, top-to-bottom below selection, bottom-to-top on top of selection 
- If horizontal overflow happens, panel is pushed to border of editable area.
- If horizontal overflow happens, and panel is larger than editable area, we are positioning panel to start of editable area. This triggers editor's own mechanisms for overflow (line breaking of toolbars etc).
- There is a minor bug with the multiple-line toolbar, when it first renders. The position is slightly wrong, you can see this in gif. This is because we are first setting position and then ckeditor breaks lines. With second and consequent renders, we are setting position with correct panel height. I think we can ignore this issue. The benefit of letting ckeditor do default overflow is too big. If this becomes a noticeable bug, we can try to find a workaround in a separate task.
- We do not offer option to ignore editable area. We can add this option, if we want to, in a separate task.
- We are displaying triangle, but in a limited capacity. It is only on top or bottom, always on canter of panel. This is the same behavior as Alloy Editor. It looks a bit broken in case of overflow. We could update it to point to selection in a future PR, but there is no direct API to it, so it could be a bit complicated.
- I did some minor SF, notably removing direction constants from global scope.  

@julien @diegonvs @dsanz 

Before PR:
before

After PR:
after

### CODE REVIEW

File: modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/_diffs/plugins/ballooneditor/plugin.js [hunk 1]
```diff
@@ -69,12 +71,12 @@
 			CKEDITOR.ui.balloonPanel.DEFAULT_PANEL_PADDING = 14;
 
 			const getSelectionDirection = (selection) => {
-				let direction = CKEDITOR.SELECTION_TOP_TO_BOTTOM;
+				let direction = SELECTION_DIRECTION.TOP_TO_BOTTOM;
```
Comment (diegonvs) [inline modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/_diffs/plugins/ballooneditor/plugin.js:74]: Nice!  🎉 

We should avoid polluting CKEditor global. When seeing this LOC at the first time I tried to search on CKEDITOR docs and I didn't found anything.

File: modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/_diffs/plugins/ballooneditor/plugin.js [hunk 2]
```diff
@@ -180,26 +196,38 @@
...
+				const editable = this.editor.editable();
+
+				const editableRect = editable.getClientRect(true);
+
+				if (editableRect.width < panelRect.width + padding) {
+					x = editableRect.x + padding;
 				}
-				if (x + panelClientRect.width > editableClientRect.width) {
-					x -= x + panelClientRect.width - editableClientRect.width;
+				else if (x < editableRect.x) {
```
Comment (diegonvs) [inline modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/_diffs/plugins/ballooneditor/plugin.js:213]: Did you changed these formulas? Or is just a weird diff?

**Review by kresimir-coko — APPROVED** [review]
LNHTM 👍

File: modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/_diffs/plugins/ballooneditor/plugin.js (hunk 2 above)
Comment (markocikos) [reply modules/apps/frontend-editor/frontend-editor-ckeditor-web/src/main/resources/META-INF/resources/_diffs/plugins/ballooneditor/plugin.js:213]: I changed them, in some cases significantly.

**Comment by markocikos** [comment]
This PR is very safe to forward. The changes are only applied to a sample module.

**Comment by markocikos** [comment]
ci:forward

**Pull request closed without merging**